src/upload.ts

```diff
@@ -0,0 +1,150 @@
+import {
+  Instrumentation, IPayload, ITotalByteLimitExceededEventState, IUploadInfo, IXhrErrorEventState,
+  State, UploadCallback
+} from "../clarity";
+import { config } from "./config";
+import { instrument, state, teardown } from "./core";
+import { debug, getCookie, getEventId, guid, isNumber, mapProperties, setCookie } from "./utils";
+
+// Counters
+let sentBytesCount: number;
+let uploadCount: number;
+
+// Storage for payloads that are compressed and are ready to be sent, but are waiting for Clarity trigger.
+// Once trigger is fired, all payloads from this queue will be sent in the order they were generated.
+let uploadQueue: IUploadInfo[];
+
+// Map from a sequence number to an UploadInfo object for payloads that were attempted to be sent, but were not succesfully delivered yet
+let uploadInfos: { [key: number]: IUploadInfo };
+
+let reUploadQueue: number[];
+
+export function upload(compressed: string, raw: IPayload, onSuccessExtra?: UploadCallback, onFailureExtra?: UploadCallback) {
+  let uploadHandler = config.uploadHandler || defaultUpload;
+  let sequenceNo = raw.envelope.sequenceNumber;
+  let onSuccess = (status: number) => {
+    onSuccessDefault(sequenceNo, status);
+    if (onSuccessExtra) {
+      onSuccessExtra(status);
+    }
+  };
+  let onFailure = (status: number) => {
+    onFailureDefault(sequenceNo, status);
+    if (onFailureExtra) {
+      onFailureExtra(status);
+    }
+  };
+  uploadInfos[sequenceNo] = { compressed, raw, failureCount: 0 };
+  uploadHandler(compressed, onSuccess, onFailure);
+
+  if (config.debug) {
+    let rawLength = JSON.stringify(raw).length;
+    debug(`** Upload #${uploadCount}: Batch #${sequenceNo}`
+          + ` ${Math.round(compressed.length / 1024.0)}KB`
+          + ` (${Math.round(rawLength / 1024.0)}KB Raw). **`);
+  }
+  uploadCount++;
+}
+
+export function enqueueUpload(compressed: string, raw: IPayload) {
+  let payloadInfo: IUploadInfo = { compressed, raw, failureCount: 0 };
+  uploadQueue.push(payloadInfo);
+}
+
+export function flushUploadQueue() {
+  uploadSequentially(uploadQueue);
+}
+
+export function uploadSequentially(uploads: IUploadInfo[]) {
```

Comment:
>uploadSequentially [](start = 16, length = 18)

this is potentially a naming nit, but with the way you handle errors you aren't actually going in sequence - if 2 fails but passes on retry you'd upload:

1-3-2

If the order of uploads isn't important, then this should just be renamed to indicate it is serial rather than sequential.

Reply:
I'm also nitty, yet so bad with variable names. I'll rename to uploadMultipleBatches to avoid confusion

---
In reply to: [168000793](https://github.com/Microsoft/clarity-js/pull/63#discussion_r168000793) [](ancestors = 168000793)